### ISSUE TITLE
BC breaks with deprecations

### ISSUE THREAD

**Issue by dereuromark** [cakephp/cakephp#11572]
I think we discussed this a while back, but it just took me by surprise again in a few cases.
If we do not follow the inheritance principle here closely, we will have soon a lot of broken user apps due to this issue.

In my case I have a contact form and a validator:
```
	/**
	 * @param \Cake\Validation\Validator $validator
	 * @return \Cake\Validation\Validator
	 */
	protected function _buildValidator(Validator $validator) { ... }
```
With 3.6 this would now break silently and can be a security issue (no validation anymore at all for all forms!) as the core code internally shortcuts it due to the deprecation refactor:
```
     * @deprecated 3.6.0 Use Form::getValidator()/setValidator() instead.
     */
    public function validator(Validator $validator = null)
    {
        deprecationWarning(
            'Form::validator() is deprecated. ' .
            'Use Form::getValidator()/setValidator() instead.'
        );

        if ($validator === null && empty($this->_validator)) {
            $validator = $this->_buildValidator(new $this->_validatorClass);
        }
        if ($validator) {
            $this->_validator = $validator;
            $this->setValidator('default', $validator);
        }

        return $this->getValidator();
    }
```
But the actual code is never calling this old one which would set the ones as CakePHP 3.x so far depended on.
It directly goes to getValidator().

## So to sum things up
We must always call internally the old methods still and only from the outside of core are allowed to replace our own old method calls with new ones. Even in plugins this is already quite dangerous and can lead to side effects in apps depending on the plugins. But this is sure worse with the core code here and apps depending on the functionality behind protected methods being overwritten as designed to be done.

Not sure how we can still have deprecation warnings then, though. How to distinguish outside vs inside usage.

There are lots of other places where core-internal using the new methods will shortcut the user land methods. We should probably revert those where it could be also critical in terms of functionality BC breaks.

**Comment by chinpei215** [comment]
I warned you about the problem in the past: https://github.com/cakephp/cakephp/issues/7954#issuecomment-172410439

Since we decided to split settters/getters, of course users' code would break if they have overridden those methods. We could mark those methods as `final`, if we want to prevent deprecated methods from overriding.

**Comment by markstory** [comment]
Protected methods aren't normally part of our [backwards compatibility](https://book.cakephp.org/3.0/en/contributing/backwards-compatibility.html) as it makes very difficult for us to make any changes as we don't use private methods.

If we want to extend backwards compatibility to protected methods we will need to start using private methods.

**Comment by markstory** [comment]
In this scenario we should document the now unused protected method in the migration guide.

**Comment by chinpei215** [comment]
Oh did we change the interface of `Form`? That feels like more than separating combined setter/sgetters. I think we should revert the change and Cookbook. Currently, our Cookbook lies as 3.5 doesn't have `validationDefault()` method.

**Comment by chinpei215** [comment]
Or we could call `_buildValidator()` in `buildValidator()` or `validationDefault()`?

**Comment by markstory** [comment]
@chinpei215 That sounds like it could it work well. The docs changes likely ended up in the wrong branch by accident.

**Comment by dereuromark** [comment]
Good thing that I accidentally was still on a switched 3.6 branch instead of 3.5 when doing tests for my app.

**Comment by ADmad** [comment]
> Currently, Cookbook lies as 3.5 doesn't have validationDefault() method.

Which page are you talking about? https://book.cakephp.org/3.0/en/core-libraries/form.html has no mention of `validationDefault()`.

**Comment by chinpei215** [comment]
@ADmad Sorry, I was reading the page about FormHelper 😛
